src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs

```diff
@@ -77,14 +79,99 @@ private async Task<Document> UseCollectionDirectly(Document document, TextSpan s
             }
 
             var generator = SyntaxGenerator.GetGenerator(document);
-            var indexNode = generator.LiteralExpression(0);
-            var elementAccessNode = generator.ElementAccessExpression(collectionSyntax.WithoutTrailingTrivia(), indexNode)
-                .WithTrailingTrivia(invocationNode.GetTrailingTrivia());
 
-            var newRoot = root.ReplaceNode(invocationNode, elementAccessNode);
+            var elementAccessNode = GetReplacementNode(methodName, generator, collectionSyntax, semanticModel);
+            if (elementAccessNode == null)
+            {
+                return document;
+            }
+
+            var newRoot = root.ReplaceNode(invocationNode, elementAccessNode.WithTrailingTrivia(invocationNode.GetTrailingTrivia()));
             return document.WithSyntaxRoot(newRoot);
         }
 
+        private static SyntaxNode GetReplacementNode(string methodName, SyntaxGenerator generator, SyntaxNode collectionSyntax, SemanticModel semanticModel)
+        {
+            var collectionSyntaxNoTrailingTrivia = collectionSyntax.WithoutTrailingTrivia();
+
+            if (methodName == "First")
+            {
+                var zeroLiteral = generator.LiteralExpression(0);
+                return generator.ElementAccessExpression(collectionSyntaxNoTrailingTrivia, zeroLiteral);
+            }
+
+            if (methodName == "Last")
+            {
+                // TODO: Handle C# 8 index expression (and vb.net equivalent if any)
+
+                if (!HasCountProperty(collectionSyntax, semanticModel))
+                {
+                    return null;
+                }
+
+                // TODO: Handle cases were `collectionSyntax` is an invocation. We would need to create some intermediate variable.
+                var countMemberAccess = generator.MemberAccessExpression(collectionSyntaxNoTrailingTrivia, "Count");
+                var oneLiteral = generator.LiteralExpression(1);
+
+                // The SubstractExpression method will wrap left and right in parenthesis but those will be automatically removed later on
+                var substraction = generator.SubtractExpression(countMemberAccess, oneLiteral);
+                return generator.ElementAccessExpression(collectionSyntaxNoTrailingTrivia, substraction);
+            }
+
+            if (methodName == "Count")
+            {
+                return HasCountProperty(collectionSyntax, semanticModel)
+                    ? generator.MemberAccessExpression(collectionSyntaxNoTrailingTrivia, "Count")
+                    : null;
+            }
+
+            Debug.Fail($"Unexpected method name '{methodName}' for {DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectlyAnalyzer.RuleId} code fix.");
+            return null;
+        }
+
+        private static bool HasCountProperty(SyntaxNode node, SemanticModel model)
```

Comment:
@mavasani I have crafted a method to find the `Count` property but you might be able to redirect me to some helper method that would be better than what I did.

Reply:
We have an analyzer that detects invocations of Linq method `Count()` and flags them if they can be replaced with property reference `.Length` or `.Count`: https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.NetCore.Analyzers/Core/Performance/UsePropertyInsteadOfCountMethodWhenAvailable.cs#L301. You can possibly move out the logic to a shared helper?

Reply:
@mavasani it seems that the other analyzer you are referring to doesn't handle `IReadOnlyCollection.Count`. Shall I create a separate PR for this and create the shared helper there?

Reply:
A separate PR sounds reasonable. Once that PR is merged in, you can update this PR to use the helper.

Reply:
@Evangelink Can you remove this private method and use the new helper method for presence of any Count property that you added in a separate PR?

Reply:
@mavasani done